src/metabase/server/middleware/session.clj

```diff
@@ -305,3 +307,89 @@
+  :type       :json
+  :default    nil)
+
+(defn session-timeout->seconds
+  "Convert a session timeout setting to seconds."
+  [{:keys [unit amount]}]
+  (when amount
+    (-> (case unit
+          "seconds" amount
+          "minutes" (* amount 60)
+          "hours"  (* amount 3600))
+        (max 60)))) ; Ensure a minimum of 60 seconds so a user can't lock themselves out
+
+(defn response-with-session-timeout-cookie
+  "Adds a cookie to the response that expires after `session-timeout-seconds` seconds."
+  [request request-time timeout-seconds response]
+  (cond
+    ;; If there is no session cookie because the user is logged out, clear the timeout cookie
+    (and (get-in request [:cookies metabase-session-timeout-cookie :value])
+         (not (or (get-in request [:cookies metabase-session-cookie :value])
+                  (get-in response [:cookies metabase-session-cookie :value]))))
+    (-> response
+        (wrap-body-if-needed)
+        (clear-cookie metabase-session-timeout-cookie))
+
+    ;; If the timeout-seconds is nil, sessions last indefinitely
+    (nil? timeout-seconds)
+    (let [cookie-options (merge
+                          {:path    "/"}
+                          (when (request.u/https? request)
+                            ;; SameSite=None is required for cross-domain full-app embedding.
+                            ;; Note that most browsers will only accept SameSite=None with secure cookies,
+                            ;; so we are setting it only over HTTPS to prevent the cookie from being rejected
+                            ;; in case of same-domain embedding. See https://github.com/metabase/metabase/issues/18553
+                            {:same-site :none
+                             :secure    true}))]
+      (-> response
+          (wrap-body-if-needed)
+          (response/set-cookie metabase-session-timeout-cookie "alive" cookie-options)))
+
+    (or
+     ;; Either the session is just starting, so we need to set the timeout cookie for the first time
+     (get-in response [:cookies metabase-session-cookie :value])
+     ;; Or the cookie is already set and not expired, so we need to reset it.
+     (get-in request [:cookies metabase-session-timeout-cookie :value]))
+    (let [cookie-options (merge
+                          {:path      "/"
+                           :expires   (t/format :rfc-1123-date-time (t/plus request-time (t/seconds timeout-seconds)))}
+                          (when (request.u/https? request)
+                            {:same-site :none
+                             :secure    true}))]
+      (-> response
+          (wrap-body-if-needed)
+          (response/set-cookie metabase-session-timeout-cookie "alive" cookie-options)))
+
+    ;; Clear the session cookie if the timeout cookie expired
+    (and (not (get-in request [:cookies metabase-session-timeout-cookie :value]))
+         (get-in request [:cookies metabase-session-cookie :value]))
+    (clear-session-cookie response)
```

Comment:
If I'm reading this right, if a request lacks a timeout cookie it will log the user out. This seems like a rough way to implement this. For example, anyone who uses Metabase's API will have quite the confusing time figuring out why everything is failing when they upgrade. I'm also a fan of using the API `http POST localhost:3000/... -Cookie $COOKIE`. Now it seems if i don't continually update that cookie I'll get logged out. It seems this will be true regardless of whether I use this new feature or not.

Reply:
@calherries @dpsutton BE should not check `metabase.TIMEOUT` to decide if it needs to remove the session cookie — everyone can manually edit `metabase.TIMEOUT` and make it not expire after the `Expires` value came from server. Server should basically update `Expires`  value of `metabase.SESSION`, `metabase.TIMEOUT` on every request by `now + sessionTimeout`

Reply:
Overall, BE should keep the session expiration date and should not rely on any FE data. Otherwise, by using devtools I could extend my session indefinitely

Reply:
Loud and clear, I implemented this incorrectly. Just to confirm my understanding: metabase.TIMEOUT can be tampered with from the FE, but metabase.SESSION can't. So metabase.TIMEOUT should only be used to implement the "log out immediately when the timeout expires" feature, which needs to be triggered from the FE (no way around that as I'm aware), but metabase.SESSION will prevent users from making additional API requests from the browser after the timeout expires.

Reply:
I just answered my own question. You can change the expiry date and value of all cookies from devtools. I'm now thinking that we need to encrypt the expiry dates of the cookies in the cookie's value so a user can't set their own. What do you think?

Reply:
Conor confirmed protecting users from extending sessions with devtools is not a must-have in Notion [here](https://www.notion.so/Offer-more-control-over-session-timeouts-bc8e84e49de44de0950385fc839b8df3?d=91ef8238970b432fb5110ba3a5e19be8#19f1a19389a345e2ab5e388db1078975).